### PR TITLE
Documentation Improvement

### LINKED ISSUE #271: Improve Documentation
## Is your feature request related to a problem? Please describe

Improve Documentation.

## Describe the solution

- Separate pages for Components documentation with links to examples
- Separate page for Clients documentation with clinks to examples
- Code comments in the examples
- Separate pages for other packages like tracing, logging, etc.
- All this pages linked in the main readme which contains general information like description, code of conduct, contribution guidelines.

### PR DESCRIPTION
**Pull request by mantzas**
## Which problem is this PR solving?

Closes #271 

## Short description of the changes

- Create new pages
- Made README lean.

- [x] [Architecture](docs/Architecture.md)
- [x] [CLI](docs/other/CLI.md)
- [x] Components
  - [x] [Async](docs/components/async/Async.md)
    - [x] [Kafka](docs/components/async/Kafka.md)
    - [x] [AMQP](docs/components/async/AMQP.md)
    - [x] [AWS SQS](docs/components/async/AWSSQS.md)
  - [x] [HTTP](docs/components/HTTP.md)
  - [x] [gRPC](docs/components/gRPC.md)
- [x] Clients
- [x] Packages
  - [x] [Reliability](docs/other/Reliability.md)
  - [x] [Observability](docs/observability/Observability.md)
  - [x] [Logging](docs/observability/Logging.md)
  - [x] [Distributed Tracing](docs/observability/DistributedTracing.md)  
  - [x] [Caching](docs/other/Caching.md)
  - [x] [Encoding](docs/other/Encoding.md)
  - [x] [Errors](docs/other/Errors.md)
- [x] [Examples](docs/Examples.md)
- [x] [Code of Conduct](docs/CodeOfConduct.md)
- [x] [Contribution Guidelines](docs/ContributionGuidelines.md)
- [x] [Acknowledgments](docs/ACKNOWLEDGMENTS.md)

### CODE REVIEW

File: docs/Architecture.md [hunk 1]
```diff
@@ -0,0 +1,233 @@
+# Architecture
+
+## Service
+
+The `Service` has the role of glueing all of the above together, which are:
```
Comment (drakos74) [inline docs/Architecture.md]:
```suggestion
The `Service` has the role of gluing all of the above together, which are:
```

Comment (drakos74) [reply docs/Architecture.md]: https://www.quora.com/Which-is-correct-gluing-or-glueing

File: docs/Architecture.md [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+```go
+Decode(v interface{}) error
+```
+
+The `Response` model contains the following properties (which are provided when calling the "constructor" `NewResponse`)
+
+- Payload, which may hold a struct of type `interface{}`
+
+### Middlewares per Route
```
Comment (drakos74) [inline docs/Architecture.md]: i missed how we ended up in the routes, i would expect an HTTP section somewhere above. Or am i missing something ?

File: docs/Architecture.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+	cache cache.TTLCache
+	// age specifies the minimum and maximum amount for max-age and min-fresh header values respectively
+	// regarding the client cache-control requests in seconds
+	age age
+}
+
+func NewRouteCache(ttlCache cache.TTLCache, age Age) *RouteCache
+```
+
+#### server cache
```
Comment (drakos74) [inline docs/Architecture.md]:
```suggestion
**server cache**
```

File: docs/Architecture.md [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+- Requests where the client control header requirements cannot be met i.e. **very low max-age** or **very high min-fresh** parameters,
+will be returned to the client with a `Warning` header present in the response. 
+
+```
+Note : When a cache is used, the handler execution might be skipped.
+That implies that all generic handler functionalities MUST be delegated to a custom middleware.
+i.e. counting number of server client requests etc ... 
+```
+
+### Usage
```
Comment (drakos74) [inline docs/Architecture.md]:
```suggestion
**Usage**
```

File: docs/Architecture.md [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+```
+
+- use the cache as a middleware
+```go
+NewRouteBuilder("/", handler).
+    WithMiddlewares(NewCachingMiddleware(NewRouteCache(cc, Age{Max: 10 * time.Second}))).
+    MethodGet()
+```
+
+#### client cache-control
```
Comment (drakos74) [inline docs/Architecture.md]:
```suggestion
**client cache-control**
```

File: docs/Architecture.md [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+
+returns a new response to the client by executing the route processing function.
+NOTE : Except for cases where a `minAge` or `maxFresh` parameter has been specified in the server.
+This is again a safety mechanism to avoid 'aggressive' clients put unexpected load on the server.
+The server is responsible to cap the refresh time, BUT must respond with a `Warning` header in such a case.
+- `only-if-cached`
+
+expects any response that is found in the cache, otherwise returns an empty response
+
+#### metrics
```
Comment (drakos74) [inline docs/Architecture.md]:
```suggestion
**metrics**
```

File: docs/Architecture.md [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+this would signify that probably a cache is not a good option for the access patterns at hand.
+
+- `eviction age`
+
+The age at which the objects are evicted from the cache is a very useful indicator. 
+If the vast amount of evictions are close to the time to live setting, it would indicate a nicely working cache.
+If we find that many evictions happen before the time to live threshold, clients would be making use cache-control headers.
+ 
+
+#### cache design reference
```
Comment (drakos74) [inline docs/Architecture.md]:
```suggestion
**cache design reference**
```

File: docs/Architecture.md [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+- https://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html
+- https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.9
+
+#### improvement considerations
```
Comment (drakos74) [inline docs/Architecture.md]:
```suggestion
**improvement considerations**
```

File: docs/ContributionGuidelines.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
+# Contribution Guidelines
+
+1. **Contributor**: An issue has to be created with a problem description and possible solutions using the github template. The better the problem and solution are described, the easier for the **Curators and Others** to understand it and the faster the process. In case of a bug, steps to reproduce will help a lot.
+2. **Curators and Others**: The curators will engage in a discussion about the problem and the possible solution. Others can join the discussion to bring other solutions and insights at any point.
```
Comment (drakos74) [inline docs/ContributionGuidelines.md:4]: should we add what we discussed ? That by being a consistent contributor you become a curator ?

**Review by drakos74 — DISMISSED** [review]

File: docs/Architecture.md (hunk 2 above)
Comment (mantzas) [reply docs/Architecture.md]: documentation is still in WIP. The description contains the unfinished work.

File: docs/ContributionGuidelines.md (hunk 9 above)
Comment (mantzas) [reply docs/ContributionGuidelines.md:4]: Would you like to add this somewhere? Probably in a different section in the guidelines.

**Comment by bhiravabhatla** [comment]
Hi Team,

Was going through set up for examples. Had a small suggestion -

As we have to run docker-compose up from inside examples directory ( mounting config for prometheus from $PWD). Can we change the go run statements to -
```
go run seventh/main.go
go run first/main.go
...
```
@mantzas  @tpaschalis

**Comment by bhiravabhatla** [comment]
Hey Team,

Couple more observations while running examples. 

* Looks like there is a breaking change in localstack, all the aws mock services are exposed at single port (4566) now (https://github.com/localstack/localstack#announcements).

We have used individual port bindings in example 4 and its **failing** now. We should be binding 4566 in docker-compose and use localhost:4566 while interacting with localstack (sqs/sns). I would happy to raise a PR for this.

* examples/sixth/greeter.pb.go is not present but mentioned in README.md  in Running the examples section.

@mantzas @tpaschalis @drakos74

**Comment by bhiravabhatla** [comment]
Any Thoughts on above. Should I open an issue for the same?

 @mantzas @tpaschalis @drakos74

**Comment by mantzas** [comment]
> Any Thoughts on above. Should I open an issue for the same?
> 
> @mantzas @tpaschalis @drakos74

Nice catch, let's do this.

**Comment by bhiravabhatla** [comment]
> > Any Thoughts on above. Should I open an issue for the same?
> > @mantzas @tpaschalis @drakos74
> 
> Nice catch, let's do this.

Opened Issue https://github.com/beatlabs/patron/issues/282 for the same.

File: docs/Architecture.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
+# Architecture
+
+Patron has two basic concepts:
+
+- Component, which define a long-running task like a Server e.g. HTTP, gRPC, Kafka consumer, etc.
```
Comment (mantzas) [inline docs/Architecture.md]:
```suggestion
- Component, which define a long-running task like a server e.g. HTTP, gRPC, Kafka consumer, etc.
```

File: docs/components/HTTP.md [hunk 11]
```diff
@@ -0,0 +1,468 @@
...
+
+The HTTP component provides the functionality for creating an HTTP server exposing the relevant routes. 
+It wraps the logic and handles the boilerplate for the `net.http` go package.
+
+The way to initialise an HTTP component is through the `patron http.Builder` struct.
+```go
+// NewBuilder initiates the HTTP component builder chain.
+// The builder instantiates the component using default values for
+// HTTP Port, Alive/Ready check functions and Read/Write timeouts.
+func NewBuilder() *Builder {
```
Comment (mantzas) [inline docs/components/HTTP.md:25]: @drakos74 
Maintaining that much of details will be an issue.
For the low level techincal doc we can use the build in https://pkg.go.dev/github.com/beatlabs/patron.
What do you think?

Comment (drakos74) [reply docs/components/HTTP.md:25]: Yes, good point. Should i replace this code segment with a short example of using the builder ?

Comment (mantzas) [reply docs/components/HTTP.md:25]: yes, that would make sense.

**Review by tpaschalis — DISMISSED** [review]

**Review by tpaschalis — APPROVED** [review]

**Review by drakos74 — APPROVED** [review]

**Pull request merged**